packages/eas-cli/src/commands/channel/list.ts

```diff
@@ -88,18 +99,77 @@ export default class ChannelList extends EasCommand {
     const exp = getExpoConfig(projectDir);
     const projectId = await getProjectIdAsync(exp);
 
-    const getAllUpdateChannelForAppResult = await getAllUpdateChannelForAppAsync({
-      appId: projectId,
-    });
-    const channels = getAllUpdateChannelForAppResult.app?.byId.updateChannels;
-    if (!channels) {
-      throw new Error(`Could not find channels on project with id ${projectId}`);
-    }
+    await queryForChannelsAsync(projectId, jsonFlag);
+  }
+}
+
+type UpdateChannelObject = GetAllChannelsForAppQuery['app']['byId']['updateChannels'][0];
+
+async function queryForChannelsAsync(projectId: string, jsonFlag: boolean): Promise<void> {
+  const queryToPerformAsync = async (
+    pageSize: number,
+    offset: number
+  ): Promise<PaginatedQueryResponse<UpdateChannelObject>> => {
+    let queriedAllBranches = false;
+    let branchOffset = 0;
+    const branchLimit = 10;
+    const paginatedQuery: PaginatedQueryResponse<UpdateChannelObject> = {
+      queryResponse: [],
+      queryResponseRawLength: 0,
+    };
+    const branchesForChannels: Record<string, UpdateChannelObject['updateBranches']> = {};
+
+    // we need to paginate our paginated query. Each channel (N) can have M branches
+    // and we need to fetch all branches to populate our 'branch mapping' portion
+    // of the UI properly.
+    do {
+      const getAllUpdateChannelForAppResult = await getAllUpdateChannelForAppAsync({
+        appId: projectId,
+        channelLimit: pageSize,
+        channelOffset: offset,
+        branchLimit,
+        branchOffset,
+      });
+      const channels = getAllUpdateChannelForAppResult.app?.byId.updateChannels;
+      if (!channels) {
+        throw new Error(`Could not find channels on project with id ${projectId}`);
+      }
+      queriedAllBranches = true;
+      // save each array of branches using a record so we can lookup in O(1) later on
+      channels.forEach(channel => {
+        if (channel.updateBranches.length === branchLimit) {
+          queriedAllBranches = false;
+        }
+
+        branchesForChannels[channel.id] = [
+          ...(branchesForChannels[channel.id] ? branchesForChannels[channel.id] : []),
+          ...channel.updateBranches,
+        ];
+      });
 
+      if (!paginatedQuery.queryResponse.length) {
+        paginatedQuery.queryResponse = channels;
+      } else {
+        paginatedQuery.queryResponse.forEach(channel => {
+          channel.updateBranches = branchesForChannels[channel.id];
+        });
+      }
+
+      paginatedQuery.queryResponseRawLength = channels.length;
+      branchOffset += branchLimit + 1;
+    } while (!queriedAllBranches);
+
+    return paginatedQuery;
+  };
```

Comment:
we need to query all branches for each channel, unfortunately, to display the proper `branchMapping` table to the user.

kind of nasty, but there is no way around it client-side. if we changed how the back-end returned or sorted branches, we could just query for the first 1 or 2. That seems like a good next task.